### PR TITLE
feat(progressive onboarding): context

### PR DESCRIPTION
**Pull request by laurabeth**
This PR resolves [] 

### Description

Creates the context that will govern progressive onboarding behaviors. No UI created as yet. Up to date w/ latest behaviors in force. 

Next steps will be to implement the necessary tooltip-like UI components, then each user workflow: saves, follows and alerts.

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

### CODE REVIEW

File: src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+  })
+
+  return parse(item)
+}
+
+export const reset = async (id: string) => {
+  await AsyncStorage.removeItem(localStorageKey(id)).catch((e) =>
+    console.log("Error writing reset to local storage", e)
+  )
+}
```
Comment (dzucconi) [inline src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx:200]: I understand that most of this is copied out of Force and that these steps will almost certainly change. But I would strongly suggest adding some spec coverage to this. [The specs there](https://github.com/artsy/force/blob/main/src/Components/ProgressiveOnboarding/__tests__/ProgressiveOnboardingContext.jest.tsx) should provide a decent starting point.

File: src/app/utils/hooks/useDidMount.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+export const useDidMount = () => {
+  const isMountedRef = useRef(false)
+  useEffect(() => {
+    isMountedRef.current = true
+    return () => {
+      isMountedRef.current = false
+    }
+  }, [])
+  return isMountedRef
+}
```
Comment (dzucconi) [inline src/app/utils/hooks/useDidMount.ts:12]: Why add this hook?

**Review by dzucconi — COMMENTED** [review]
I don't really have a problem with this, but I'd probably approach this the other way around: get a single simple tooltip displaying first before trying to predict what you might need to support it.

File: src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+const ProgressiveOnboardingContext = createContext<{
+  dismissed: DismissedKey[]
+  dismiss: (key: ProgressiveOnboardingKey | ProgressiveOnboardingKey[]) => void
+  isDismissed: (key: ProgressiveOnboardingKey) => DismissedKeyStatus
+}>({
+  dismissed: [],
+  dismiss: () => {},
+  isDismissed: (_key) => ({ status: false, timestamp: 0 }),
+})
```
Comment (damassi) [inline src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx:27]: This is non-blocking, but worth noting that in Eigen we use easy-peasy, and there's a much nicer way to manage local contexts should you wish to use it:

https://easy-peasy.vercel.app/docs/api/create-context-store.html

I would suggest going this route, but can also understand not wanting to refactor.

File: src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+  keys.forEach(async (key) => {
+    const item = await AsyncStorage.getItem(localStorageKey(id))
+    const dismissed = parse(item)
+
+    await AsyncStorage.setItem(
+      localStorageKey(id),
+      JSON.stringify(uniqBy([...dismissed, { key, timestamp }], ({ key }) => key))
+    )
+  })
```
Comment (MounirDhahri) [inline src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx:184]: **thought:** I see you are here trying to set a value for the same key (`localStorageKet(id)`) in AsyncStorage many times in sequence. Is it possible to do that only one time when you know the final value?

**Review by damassi — CHANGES_REQUESTED** [review]
We should be using easy-peasy for the context, per @MounirDhahri's note about out of the box persistence, etc. It will simplify this up and down.

File: src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx (hunk 3 above)
Comment (laurabeth) [reply src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingContext.tsx:27]: Hey guys! Stepped away from this for future friday, but I'll be back to it shortly. Great feedback and I don't mind the refactor at all now that I've had a little breathing room from the work.

Thanks for taking a look.

**Pull request closed without merging**